src/common/base/CollectNSucceeded-inl.h

```diff
@@ -11,59 +11,109 @@ namespace nebula {
+        : eval(std::forward<ResultEval>(e))
+        , results(total)
+        , nTotal(total) {}
 
     ResultEval eval;
-    Result results;
-    std::atomic<size_t> numCompleted = {0};
+    std::vector<folly::Optional<folly::Try<FutureReturnType<FutureIter>>>> results;
+    std::atomic<size_t> nCompleted = {0};
     std::atomic<size_t> nSucceeded = {0};
+    std::atomic<bool> fulfilled = false;
     folly::Promise<Result> promise;
-    size_t nTotal;
+    const size_t nTotal;
   };
 
   size_t total = size_t(std::distance(first, last));
   DCHECK_GE(total, 0U);
 
   if (total < n) {
-    return folly::Future<Result>(
+    return folly::SemiFuture<Result>(
         folly::exception_wrapper(std::runtime_error("Not enough futures")));
   }
 
-  auto ctx = std::make_shared<Context>(total, std::forward<ResultEval>(eval));
+  std::vector<folly::futures::detail::DeferredWrapper> executors;
+  folly::futures::detail::stealDeferredExecutors(executors, first, last);
 
   // for each succeeded Future, add to the result list, until
   // we have required number of futures, at which point we fulfil
   // the promise with the result list
+  auto ctx = std::make_shared<Context>(total, std::forward<ResultEval>(eval));
   for (size_t index = 0; first != last; ++first, ++index) {
-    first->setCallback_([n, ctx, index](auto, folly::Try<FutureReturnType<FutureIter>>&& t) {
-      if (!ctx->promise.isFulfilled()) {
+    first->setCallback_([n, ctx, index](auto&&,
+                                        folly::Try<FutureReturnType<FutureIter>>&& t) {
+      VLOG(3) << "Received the callback from the future " << index;
+
+      // relaxed because this guards control but does not guard data
+      auto const c = 1 + ctx->nCompleted.fetch_add(1, std::memory_order_relaxed);
+      auto s = ctx->nSucceeded.load();
+
+      if (!ctx->fulfilled.load()) {
         if (!t.hasException()) {
           if (ctx->eval(index, t.value())) {
-            ++ctx->nSucceeded;
+            // release because the stored values in all threads must be visible below
+            // acquire because no stored value is permitted to be fetched early
+            s = 1 + ctx->nSucceeded.fetch_add(1, std::memory_order_acq_rel);
           }
-          ctx->results.emplace_back(index, std::move(t.value()));
+        } else {
+          // Got exception
+          DLOG(ERROR) << "Exception receided: " << t.exception().what().toStdString();
+        }
+
+        ctx->results[index] = std::move(t);
+        if (c < ctx->nTotal && s < n) {
+          return;
         }
-        if ((++ctx->numCompleted) == ctx->nTotal || ctx->nSucceeded == n) {
-          // Done
-          VLOG(2) << "Set Value [completed=" << ctx->numCompleted << ", total=" << ctx->nTotal
-                  << ", Result list size=" << ctx->results.size() << "]";
-          ctx->promise.setValue(std::move(ctx->results));
+
+        bool fulfilled = false;
+        if (!ctx->fulfilled.compare_exchange_strong(fulfilled, true)) {
```

Comment:
One of a big issue of `collectNSucceeded` is that it is no thread-safe, which requires all threads need to run on same executor and hurts performance quite a lot. We need to reconsider about it later, for example, use `SharedPromise`.